meeteval/wer/wer/time_constrained.py

```diff
@@ -378,87 +416,153 @@ def remove_overlaps(
-            f'{len(segments.transcript)} != {len(segments.timings)}'
-        )
-
-    for t in segments.timings:
-        if t[1] < t[0]:
-            raise ValueError(f'The end time of an interval must be larger than the start time. Found {t} in {name}')
-
+def sort_and_validate(segments: SegLST, sort, pseudo_word_level_timing, name):
+    """
+    Args:
+        segments:
+        sort: How to sort words/segments. Options:
+            - `True`: sort by segment start time and assert that the word-level timings are sorted by start time
+            - `False`: do not sort and do not check word order
+            - `'segment'`: sort segments by start time and do not check word order
+            - `'word'`: sort words by start time
+        pseudo_word_level_timing:
+        name:
+
+    >>> segments = SegLST([{'words': 'c d', 'start_time': 1, 'end_time': 3}, {'words': 'a b', 'start_time': 0, 'end_time': 3}])
+    >>> sort_and_validate(segments, True, 'character_based', 'test')
+    Traceback (most recent call last):
+    ...
+    ValueError: The order of word-level timings contradicts the segment-level order in test: 2 of 4 times.
+    Consider setting sort to False or "segment" or "word".
+    >>> sort_and_validate(segments, False, 'character_based', 'test')
+    SegLST(segments=[{'words': 'c', 'start_time': 1.0, 'end_time': 2.0}, {'words': 'd', 'start_time': 2.0, 'end_time': 3.0}, {'words': 'a', 'start_time': 0.0, 'end_time': 1.5}, {'words': 'b', 'start_time': 1.5, 'end_time': 3.0}])
+    >>> sort_and_validate(segments, 'segment', 'character_based', 'test')
+    SegLST(segments=[{'words': 'a', 'start_time': 0.0, 'end_time': 1.5}, {'words': 'b', 'start_time': 1.5, 'end_time': 3.0}, {'words': 'c', 'start_time': 1.0, 'end_time': 2.0}, {'words': 'd', 'start_time': 2.0, 'end_time': 3.0}])
+    >>> sort_and_validate(segments, 'word', 'character_based', 'test')
+    SegLST(segments=[{'words': 'a', 'start_time': 0.0, 'end_time': 1.5}, {'words': 'c', 'start_time': 1.0, 'end_time': 2.0}, {'words': 'b', 'start_time': 1.5, 'end_time': 3.0}, {'words': 'd', 'start_time': 2.0, 'end_time': 3.0}])
+    """
     if sort not in (True, False, 'segment', 'word'):
         raise ValueError(f'Invalid value for sort: {sort}. Choose one of True, False, "segment", "word"')
 
+    for s in segments:
+        if s['end_time'] < s['start_time']:
+            raise ValueError(f'The end time of an interval must be larger than the start time. Found {s} in {name}')
+
     if sort in (True, 'segment', 'word'):
-        segments = sort_segments(segments)
+        segments = segments.sorted('start_time')
 
     words = get_pseudo_word_level_timings(segments, pseudo_word_level_timing)
 
     # Check whether words are sorted by start time
-    words_sorted = sort_segments(words)
-    prune = True
+    words_sorted = words.sorted('start_time')
+    # TODO: only check relevant keys? That would speed things up if the user provides a lot of custom keys
```

Comment:
That is not necessary:
```
class Foo:
    def __init__(self, obj):
        self.obj = obj
    def __eq__(self, other):
        print('__eq__')
        return self.obj == other.obj

a = [Foo('a'), Foo('b')]
b = [Foo('a'), Foo('b')]

print(a == a)  # True
print(a == b)  # __eq__ __eq__ True
```
The identity of the objects should be preserved by sort and hence the comparison is cheap, if they are the same.

You may want to add a comment and change the list comprehension:
```
contradictions = [not(a is b or a == b) for a, b in zip(words_sorted, words)]
```

Reply:
Thanks, I missed that. The critical point is the  `sorted_words != words` line, but I the same thing should hold there. I'll just remove the TODO comment

Reply:
The `words_sorted != words` will be cheap, because python first tests the identity and in most cases it will be True. So that should be fine.

A problem could be the `a != b`, when a warning is raised (error is fine, that one should never be caught).
In that case, we could hit the case that you mentioned.

You could leave there a comment. I was not sure, whether Python uses first an identity check on the elements of a list. Depending on the outer code, this could change in the future (identity fails, while equality holds).

Reply:
Python's sort doesn't keep identity when the list is already sorted, it only keeps the identity of the elements:
```py
l = [1,2,3]
sorted(l) is l
Out[3]: False
sorted(l) == l
Out[4]: True

l = [{'k': 0}, {'k': 1}]
sorted(l, key=lambda x: x['k'])[0] is l[0]
Out[8]: True
```
But that shouldn't hurt much either. I'll add a comment

Reply:
Yes, that was what I expected: Python shouldn't modify the entries.
So whatever we do, if the order is not changed, the default python implementation has the best performance.